packages/financial-templates-lib/src/clients/InsuredBridgeL1Client.ts

```diff
@@ -0,0 +1,181 @@
+        whitelistedTokenEvent.returnValues.bridgePool
+      ) as unknown) as BridgePoolWeb3;
+      this.relays[l1Token] = {};
+    }
+  }
+
+  async update(): Promise<void> {
+    // Define a config to bound the queries by.
+    const blockSearchConfig = {
+      fromBlock: this.firstBlockToSearch,
+      toBlock: this.endingBlockNumber || (await this.l1Web3.eth.getBlockNumber()),
+    };
+
+    if (!this._isClientInitialized()) await this._initialSetup(blockSearchConfig.fromBlock);
+
+    // Fetch event information
+    // TODO: consider optimizing this further. Right now it will make a series of sequential BlueBird calls for each pool.
+    for (const l1Token of Object.keys(this.bridgePools)) {
+      const [
+        depositRelayedEvents,
+        relaySpedUpEvents,
+        relayDisputedEvents,
+        relaySettledEvents,
+      ] = await BluebirdPromise.all([
+        this.bridgePools[l1Token].getPastEvents("DepositRelayed", blockSearchConfig),
+        this.bridgePools[l1Token].getPastEvents("RelaySpedUp", blockSearchConfig),
+        this.bridgePools[l1Token].getPastEvents("RelayDisputed", blockSearchConfig),
+        this.bridgePools[l1Token].getPastEvents("RelaySettled", blockSearchConfig),
+      ]);
+
+      for (const depositRelayedEvent of depositRelayedEvents) {
+        const relayData = {
+          depositId: Number(depositRelayedEvent.returnValues.depositId),
+          sender: depositRelayedEvent.returnValues.sender,
+          slowRelayer: depositRelayedEvent.returnValues.slowRelayer,
+          disputedSlowRelayers: [],
+          instantRelayer: ZERO_ADDRESS,
+          depositTimestamp: depositRelayedEvent.returnValues.depositTimestamp,
+          recipient: depositRelayedEvent.returnValues.recipient,
+          l1Token: depositRelayedEvent.returnValues.l1Token,
+          amount: depositRelayedEvent.returnValues.amount,
+          slowRelayFeePct: depositRelayedEvent.returnValues.slowRelayFeePct,
+          instantRelayFeePct: depositRelayedEvent.returnValues.instantRelayFeePct,
+          realizedLpFeePct: depositRelayedEvent.returnValues.realizedLpFeePct,
+          priceRequestAncillaryDataHash: depositRelayedEvent.returnValues.priceRequestAncillaryDataHash,
+          depositHash: depositRelayedEvent.returnValues.depositHash,
+          depositContract: depositRelayedEvent.returnValues.depositContract,
+          relayState: relayState.Pending,
+        };
+
+        // If the local data contains this deposit ID then this is a re-relay of a disputed relay. In this case, we need
+        // to update the data accordingly as well as store the previous slow relayers.
+        if (this.relays[l1Token][relayData.depositHash]) {
+          // Bring the previous slow relayer from the data from the previous run and store it.
+          const previousSlowRelayer = this.relays[l1Token][relayData.depositHash].slowRelayer;
+          this.relays[l1Token][relayData.depositHash].disputedSlowRelayers.push(previousSlowRelayer);
```

Comment:
Do you anticipate it being useful to know the prior `disputedSlowRelayers` for a given relay?

Reply:
it _might be_. happy to remove. if we don't need the data. I thought there could be a case where the bot needs to know this when settling disputes but I don't think it _needs_ this data. I was rather trying to build up complete information about a given request. without this you can't tell how many times a relay has been re-relayed and disputed (if that was to happen multiple times).

Reply:
> without this you can't tell how many times a relay has been re-relayed and disputed (if that was to happen multiple times).

This is true, but I'm not sure its useful. We can keep for now but maybe add a TODO to remove it if not used. To call `settleRelay` you only need the `DepositData`